### DISCUSSION TITLE
Comparing MMT system with that of KH

### DISCUSSION THREAD

**Discussion by hmaskat17** [ec2u/data#26] [Q&A] [answered]
We discussed with @juhovainio that it would be nice to be able to compare the events we have in our MMT system with Knowledge Hub's events.

This comparison would give us more insight on the status and quality of our data. It would, for example, give us clues to whether there has been problems with importing the KH data into our systems and we would be aware of these issues in advance.

Currently, the KH API returns (presumably) a total database count when making requests to the root url `https://data.ec2u.eu/` - which is nice to have. **Does this count include events that have already passed (that is, events whose endDate is greater than the current date)?** 

If this count only includes current events, then we should be OK with comparing event count in our MMT system with the count the root url returns. At least this would give us an overview of the situation and we could, for example, build an alert system when the counts deviate from each other.

If the count includes events that have already passed, then we would propose that either it contains only current events or, as an example, that the API is updated to support query parameters for the endDate property when making requests along with returning the total count of events it finds.

In short, we want to be able to compare the events in our system with that of Knowledge Hub. How could this be done?

@knoan

**Comment by knoan** [comment, marked as answer]
> Currently, the KH API returns (presumably) a total database count when making requests to the root url https://data.ec2u.eu/ - which is nice to have.

The current root API is provisional and cannot be relied upon for future monitoring; however, the https://data.ec2u.eu/events/ port already supports extensive [faceted search](https://metreeca.github.io/java/reference/faceted-search) facilities that could easily provide you with selective stats and breakdowns, e.g.

* general count / `GET https://data.ec2u.eu/events/?.stats`
* breakdown by university / `GET https://data.ec2u.eu/events/?.terms=university`
* breakdown by publisher / `GET https://data.ec2u.eu/events/?.terms=publisher`

Both query types may be filtered by adding the following parameters to the base URL:

* by last update / `GET https://data.ec2u.eu/events/?.stats&>modified=2022-06-01T00:00:00Z`
* by end date /  `GET https://data.ec2u.eu/events/?.stats&>endDate=2022-06-01T00:00:00Z`

Breakdowns/filters may be actually built on any exposed event property, but should you need more specific queries, we could easily arrange that with custom URLs.

> Does this count include events that have already passed (that is, events whose endDate is greater than the current date)?

Stale events are expected to be purged from the graph nightly around 00:00 UTC, if at least one of the following conditions is true:

* the latest between `startDate` and `endDate` is earlier than the current date
* neither `startDate` or `endDate` is provided and the `modified` timestamp is earlier that 30 days ago
